system/lib/wasmfs/js_api.cpp

```diff
@@ -188,4 +190,44 @@ void _wasmfs_readdir_finish(struct wasmfs_readdir_state* state) {
   free(state);
 }
 
+// Given a path, parse in into a full absolute path relative to the current
+// directory. Returns an allocated buffer that the caller must free, or 0 on
+// error.
+char* _wasmfs_parse_path(char* path) {
+  auto parsed = path::parseFile((char*)path);
+  if (auto err = parsed.getError()) {
+    return 0;
+  }
+  auto& file = parsed.getFile();
+
+  // Traverse up and build the full path.
+  std::vector<std::string> chunks;
+  while (1) {
+    auto parent = file->locked().getParent();
+    if (parent == file) {
+      break;
+    }
+    chunks.push_back(parent->locked().getName(file));
+    file = parent;
+  }
+
+  std::string ret;
+  for (auto it = chunks.rbegin(); it != chunks.rend(); ++it) {
```

Comment:
Why can't you use `for (const auto& chunk: chunks) {` here?

Reply:
It has to scan them in reverse order. Note `rbegin` and not `begin` etc.

Reply:
Oh I see. .that is unfortunate.. I see overt usage of iterators like this a code smell now but perhaps I should reevaluate.

Is that because the `chunks` are stored in the reverse order starting the with leaf node of the path?  Perhaps we should build up the list in the right order above using `insert(0)` instead of push_back?   I think it would make the code more readable, would maybe cost a little more.   

Since paths won't have many components perhaps a dqueue would work well here instead of vector in reverse order?

Reply:
In C++20 there is a nice way to do this:
```cpp
for (auto x : std::views::reverse(y)) {
  ...
}
```
We are on c++17 for WasmFS, but maybe we could upgrade soon..? It's only ever compiled with our own clang. So once LLVM supports that, we can use it.

`insert(0)` would be slow for deep paths. I think iterating in reverse is the right thing as it's the most efficient.

Reply:
Is `std::dqueue + push_front` better than iterating on `views::reverse`? I think the iteration is much nicer (once we can get c++20, but as I said, I think that could be close).. Iteration is also more memory efficient since it allocates less.